Implements maliput::dragway::Lane::ToLanePosition().

liangfok opened the pull request:
This is now possible given #5302, which fixes the API contract for this method.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5346)
<!-- Reviewable:end -->

liangfok commented:
+@jadecastro for feature review please.
+@sammy-tri and +@soonho-tri for platform review please.
+@jwnimmer-tri for car review please.

Thanks!

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-:-KeApHKNN59gPvRYoaH_:bvni7as)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> for automotive design review.

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/maliput/dragway/lane.cc, line 10 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-KeBXy9e2NiofWRUmcEr:-KeBXy9e2NiofWRUmcEs:b-ciom1o) ([raw file](https://github.com/robotlocomotion/drake/blob/25e4c83d0d87874ab9c8d4f762a4d6a78f7ba6a6/drake/automotive/maliput/dragway/lane.cc#L10)):*
> ```cpp
> #include "drake/automotive/maliput/dragway/segment.h"
> #include "drake/common/drake_assert.h"
> #include "drake/common/text_logging.h"
> ```

Unused?

---

*[drake/automotive/maliput/dragway/lane.cc, line 25 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-KeBXiuI1hV43fljDhQH:-KeBXiuI1hV43fljDhQI:b-qp7owj) ([raw file](https://github.com/robotlocomotion/drake/blob/25e4c83d0d87874ab9c8d4f762a4d6a78f7ba6a6/drake/automotive/maliput/dragway/lane.cc#L25)):*
> ```cpp
> // TODO(liang.fok) Once c++17 or later is used, switch to std::clamp().
> //
> double clamp(double value, double min, double max) {
> ```

This is now the third copy of this function.  (I have one in simple_car also.)  Please add a TODO for us to move it into a common location.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-:-KeBYCQr7koxfstCQADs:bi3brdl)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Pass complete. PTAL.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 439 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-KeBuKVe_kAqQER9Fa6w:-KeBuKVfoeLbyzjpcU-u:bohqmri) ([raw file](https://github.com/robotlocomotion/drake/blob/25e4c83d0d87874ab9c8d4f762a4d6a78f7ba6a6/drake/automotive/maliput/dragway/test/dragway_test.cc#L439)):*
> ```cpp
> }
> 
> // Tests dragway::Lane::ToLanePosition() using geographic positions that whose
> ```

BTW, nix "that".

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 485 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-KeBnWU6XmI_BxHQcjrS:-KeBnWU6XmI_BxHQcjrT:b-sfoidm) ([raw file](https://github.com/robotlocomotion/drake/blob/25e4c83d0d87874ab9c8d4f762a4d6a78f7ba6a6/drake/automotive/maliput/dragway/test/dragway_test.cc#L485)):*
> ```cpp
>   const double max_x = length_;
>   const double min_y = -lane_width_ / 2 - shoulder_width_;
>   const double max_y = lane_width_ / 2+ shoulder_width_;
> ```

BTW, add a space between the `2` and the `+`.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 490 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-KeBnGYNp4OWkDZee9Qs:-KeBnGYOsQRJG0PNjoI9:b-b4c87y) ([raw file](https://github.com/robotlocomotion/drake/blob/25e4c83d0d87874ab9c8d4f762a4d6a78f7ba6a6/drake/automotive/maliput/dragway/test/dragway_test.cc#L490)):*
> ```cpp
>   // the driveable area are evaluated.
>   const std::vector<double> x_test_cases{
>     min_x - 10, min_x - 1e-10, min_x, max_x, max_x + 1e10, max_x + 10};
> ```

Typo? `1e-10`. Ditto below.
(very different numbers!)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-:-KeBuhSekVURf5u7q_ys:b-wjpjkm)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 386 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-KeBwtt4ppzF3wwftjkD:-KeBwtt4ppzF3wwftjkE:bdg7nbm) ([raw file](https://github.com/robotlocomotion/drake/blob/25e4c83d0d87874ab9c8d4f762a4d6a78f7ba6a6/drake/automotive/maliput/dragway/test/dragway_test.cc#L386)):*
> ```cpp
>   // the driveable area are evaluated.
>   const std::vector<double> x_test_cases{
>     test_x_min, x_min - 1e-10, x_max + 1e10, test_x_max};
> ```

BTW, typo? `1e-10`.  Ditto below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5346)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks.

---

Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[drake/automotive/maliput/dragway/lane.cc, line 10 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-KeBXy9e2NiofWRUmcEr:-KeBtHVryuhVQZ3qcDWt:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/25e4c83d0d87874ab9c8d4f762a4d6a78f7ba6a6/drake/automotive/maliput/dragway/lane.cc#L10)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Unused?
</blockquote></details>

Done.

---

*[drake/automotive/maliput/dragway/lane.cc, line 25 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-KeBXiuI1hV43fljDhQH:-KeBtpdIa9dsHEJZ4dn9:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/25e4c83d0d87874ab9c8d4f762a4d6a78f7ba6a6/drake/automotive/maliput/dragway/lane.cc#L25)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This is now the third copy of this function.  (I have one in simple_car also.)  Please add a TODO for us to move it into a common location.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 386 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-KeBwtt4ppzF3wwftjkD:-KeBxvxaZHgWGSzFZjKS:b-d0erg5) ([raw file](https://github.com/robotlocomotion/drake/blob/25e4c83d0d87874ab9c8d4f762a4d6a78f7ba6a6/drake/automotive/maliput/dragway/test/dragway_test.cc#L386)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW, typo? `1e-10`.  Ditto below.
</blockquote></details>

Done. Ha!! Good to know it worked regardless! :smile: 

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 439 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-KeBuKVe_kAqQER9Fa6w:-KeBy3LIamWqkaepmL5A:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/25e4c83d0d87874ab9c8d4f762a4d6a78f7ba6a6/drake/automotive/maliput/dragway/test/dragway_test.cc#L439)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW, nix "that".
</blockquote></details>

Done.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 485 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-KeBnWU6XmI_BxHQcjrS:-KeBy648G4JNyBuUiXKc:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/25e4c83d0d87874ab9c8d4f762a4d6a78f7ba6a6/drake/automotive/maliput/dragway/test/dragway_test.cc#L485)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW, add a space between the `2` and the `+`.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 490 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-KeBnGYNp4OWkDZee9Qs:-KeBy9njb7t2H64NXJjq:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/25e4c83d0d87874ab9c8d4f762a4d6a78f7ba6a6/drake/automotive/maliput/dragway/test/dragway_test.cc#L490)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Typo? `1e-10`. Ditto below.
(very different numbers!)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-:-KeBxgTwn9hL9xDKpRSe:bblnx7v)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-:-KeBzjaFvmEzoRqZG4jV:bg7igkv)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5346)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5346)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/maliput/dragway/lane.cc, line 126 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-KeElL5fJ3nNwk6h67i5:-KeElL5fJ3nNwk6h67i6:b7fy3p8) ([raw file](https://github.com/robotlocomotion/drake/blob/bfc4345714bb0c7fc3b7fc41efe037725010ae4d/drake/automotive/maliput/dragway/lane.cc#L126)):*
> ```cpp
>   const double max_y = driveable_bounds_.r_max + y_offset_;
> 
>   api::GeoPosition closest_point;
> ```

FYI, the following also works:
```c++
  const api::GeoPosition closest_point{clamp(geo_pos.x, min_x, max_x),
                                       clamp(geo_pos.y, min_y, max_y),
                                       geo_pos.z};
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-:-KeElxrcWdJXq2QD1T6L:bnfp4nl)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
-@sammy-tri I'm not going to be able to get to this in a timely manner.  It seems to have plenty of review already...

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-:-KeF1BP_ZGss4K27bWA8:b-ig6fjo)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
I'll bless this with my platform hat on.

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-:-KeF1ivgJthj_DTzFtdF:bers2xj)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/maliput/dragway/lane.cc, line 126 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-KeElL5fJ3nNwk6h67i5:-KeF5H_tHf5OPCJ7KjPH:boyexb3) ([raw file](https://github.com/robotlocomotion/drake/blob/bfc4345714bb0c7fc3b7fc41efe037725010ae4d/drake/automotive/maliput/dragway/lane.cc#L126)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, the following also works:
```c++
  const api::GeoPosition closest_point{clamp(geo_pos.x, min_x, max_x),
                                       clamp(geo_pos.y, min_y, max_y),
                                       geo_pos.z};
```
</blockquote></details>

Done. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5346)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5346)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
"LGTM missing from assignee Liang"..?

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5346#-:-KeFOeKk4TDExYcVNJxM:bu1mv7a)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Oops. Done. Thanks.
